File: packages/sdk/src/domains/rooms.ts
/**
 * The rooms domain: discover public rooms, join and leave them, and observe their membership and state.
 *
 * A room is a large, discoverable conversation. Joining one returns an ordinary {@link
 * ConversationSummary}-style handle — a `conversationId`, an {@link EncryptionMode}, and the current
 * `lastSeq` — so once joined a room is read and written through the same messaging domain as any other
 * conversation. What is distinct is the lifecycle around it: rooms are browsed and searched ({@link
 * list}), entered and left ({@link join} / {@link leave}), and they emit live streams a direct
 * chat does not.
 *
 * # The three event streams
 *
 * {@link onMember} carries per-member joins and leaves; {@link onState} carries coalesced counters and
 * metadata — online count, member count, topic, slow-mode interval — as *deltas*, so a handler applies
 * each field it receives onto the room state it already holds rather than treating an event as a full
 * snapshot. {@link onVote} carries the running tally of a kick vote, coalesced per room, so every member
 * watches the same count climb toward the threshold. All three are broadcast to every device in the
 * room, so a client sees membership churn, counter movement, and vote progress without polling. A
 * joined room's encryption is still end-to-end where the summary says so; discoverability changes who
 * may enter, not whether content is sealed.
 *
 * # Moderation
 *
 * Two paths remove trouble. {@link voteKick} is the members' own recourse: any member calls for a kick,
 * and when half the room has cast the same call the target is removed — the reply and the {@link onVote}
 * stream both report the tally. {@link sanction} is the staff path: a member with rank enough (a
 * moderator, or a global admin) mutes, kicks, or bans a lower-ranked member outright, no vote needed.
 */

import type { Id } from '@migo/wire';
import {
  OP,
  encodeRoomJoinRequest,
  decodeRoomJoinResponse,
  encodeRoomLeaveRequest,
  encodeRoomListRequest,
  decodeRoomListResponse,
  encodeRoomCreate,
  encodeRosterReq,
  decodeRosterResponse,
  decodeRoomMemberEvent,
  decodeRoomStateEvent,
  decodeRoomVoteEvent,
  encodeRoomVoteKick,
  decodeRoomVoteKickResponse,
  encodeRoomSanction,
  decodeAcknowledged,
} from '@migo/protocol';
import type {
  RoomCreate,
  RoomJoinRequest,
  RoomJoinResponse,
  RoomLeaveRequest,
  RoomListRequest,
  RoomListResponse,
  RoomMemberEvent,
  RoomStateEvent,
  RoomVoteEvent,
  RoomVoteKick,
  RoomVoteKickResponse,
  RoomSanction,
  SanctionAction,
  RosterEntry,
  RosterReq,
  RosterResponse,
  Acknowledged,
  RoomKind,
} from '@migo/protocol';

import { ListenerSet } from './listeners.js';
import type { Listener } from './listeners.js';
import type { EventErrorHandler, Rpc } from './rpc.js';

/** Optional filters for a room search. Any combination narrows the listing. */
export interface RoomListFilter {
  /** A free-text query matched against room name, topic, and description. */
  query?: string;
  /** Restrict to a category. */
  category?: string;
  /** Restrict to a language code. */
  language?: string;
  /** Restrict to a country code. */
  country?: string;
  /** The {@link RoomListResponse.nextCursor} from a previous page. */
  cursor?: string;
}

/**
 * Browse, join, leave, and observe rooms.
 *
 * One instance per client. {@link start} begins delivering the two event streams; {@link list},
 * {@link join}, and {@link leave} work independently of it.
 */
export class RoomsDomain {
  readonly #rpc: Rpc;
  readonly #memberListeners: ListenerSet<RoomMemberEvent>;
  readonly #stateListeners: ListenerSet<RoomStateEvent>;
  readonly #voteListeners: ListenerSet<RoomVoteEvent>;
  #unsubscribes: Array<() => void> = [];

  constructor(rpc: Rpc, onEventError?: EventErrorHandler) {
    this.#rpc = rpc;
    this.#memberListeners = new ListenerSet(OP.ROOM_MEMBER_EVENT, onEventError);
    this.#stateListeners = new ListenerSet(OP.ROOM_STATE_EVENT, onEventError);
    this.#voteListeners = new ListenerSet(OP.ROOM_VOTE_EVENT, onEventError);
  }

  /** Begins delivering room member, state, and vote events to registered handlers. Idempotent. */
  start(): void {
    if (this.#unsubscribes.length > 0) {
      return;
    }
    this.#unsubscribes.push(
      this.#rpc.on(OP.ROOM_MEMBER_EVENT, decodeRoomMemberEvent, (event) =>
        this.#memberListeners.deliver(event),
      ),
      this.#rpc.on(OP.ROOM_STATE_EVENT, decodeRoomStateEvent, (event) =>
        this.#stateListeners.deliver(event),
      ),
      this.#rpc.on(OP.ROOM_VOTE_EVENT, decodeRoomVoteEvent, (event) =>
        this.#voteListeners.deliver(event),
      ),
    );
  }

  /** Stops delivering room events. Registered handlers are kept for a later {@link start}. */
  stop(): void {
    for (const unsubscribe of this.#unsubscribes) {
      unsubscribe();
    }
    this.#unsubscribes = [];
  }

  /** Registers a handler for per-member join and leave events. Returns an unsubscribe function. */
  onMember(handler: Listener<RoomMemberEvent>): () => void {
    return this.#memberListeners.add(handler);
  }

  /**
   * Registers a handler for coalesced room-state deltas. Returns an unsubscribe function.
   *
   * Each event carries only the fields that changed; apply them onto held state rather than replacing
   * it.
   */
  onState(handler: Listener<RoomStateEvent>): () => void {
    return this.#stateListeners.add(handler);
  }

  /**
   * Registers a handler for kick-vote tallies. Returns an unsubscribe function.
   *
   * Each event names the target and the count so far against the threshold; `closed` marks the vote's
   * end (it passed, expired, or the target left). The stream is coalesced per room, so a handler holds
   * one tally per target and replaces it as counts arrive rather than accumulating.
   */
  onVote(handler: Listener<RoomVoteEvent>): () => void {
    return this.#voteListeners.add(handler);
  }

  /**
   * Searches discoverable rooms.
   *
   * `limit` bounds one page; pass {@link RoomListResponse.nextCursor} as {@link RoomListFilter.cursor}
   * to page. A response with no `nextCursor` is the last page.
   */
  async list(limit: number, filter: RoomListFilter = {}): Promise<RoomListResponse> {
    const request: RoomListRequest = { limit };
    if (filter.query !== undefined) {
      request.query = filter.query;
    }
    if (filter.category !== undefined) {
      request.category = filter.category;
    }
    if (filter.language !== undefined) {
      request.language = filter.language;
    }
    if (filter.country !== undefined) {
      request.country = filter.country;
    }
    if (filter.cursor !== undefined) {
      request.cursor = filter.cursor;
    }
    return this.#rpc.call(OP.ROOM_LIST, encodeRoomListRequest, decodeRoomListResponse, request);
  }

  /**
   * Creates a room and enters it, resolving with the same join handle {@link join} returns.
   *
   * The caller becomes the room's Owner ({@link RoomRole.Owner}): the one role that can appoint
   * managers and the one a room cannot lose. `slug` is the room's permanent address and must be
   * unique server-side; `kind` picks the governance line — {@link RoomKind.Public} for a
   * community room, {@link RoomKind.Managed} for one under server moderation. The reply is a
   * join response because creation is entry: the creator is the first member.
   */
  async create(
    slug: string,
    name: string,
    kind: RoomKind,
    topic?: string,
  ): Promise<RoomJoinResponse> {
    const request: RoomCreate = { slug, name, kind };
    if (topic !== undefined) {
      request.topic = topic;
    }
    return this.#rpc.call(OP.ROOM_CREATE, encodeRoomCreate, decodeRoomJoinResponse, request);
  }

  /**
   * Joins a room, optionally with an invite code for a non-public one.
   *
   * Resolves with the room summary plus the conversation handle to read and write it through the
   * messaging domain: its `conversationId`, its `encryption` mode, and the current `lastSeq` to start
   * syncing from.
   */
  async join(roomId: Id, inviteCode?: string): Promise<RoomJoinResponse> {
    const request: RoomJoinRequest = { roomId };
    if (inviteCode !== undefined) {
      request.inviteCode = inviteCode;
    }
    return this.#rpc.call(OP.ROOM_JOIN, encodeRoomJoinRequest, decodeRoomJoinResponse, request);
  }

  /**
   * Leaves a room.
   *
   * After leaving, forget the room's crypto state through the messaging domain — a room rotates its
   * sender key on membership change so departed members cannot read new messages, and the local
   * receiver state for it is no longer useful.
   */
  async leave(roomId: Id): Promise<Acknowledged> {
    const request: RoomLeaveRequest = { roomId };
    return this.#rpc.call(OP.ROOM_LEAVE, encodeRoomLeaveRequest, decodeAcknowledged, request);
  }

  /**
   * Casts a vote to kick a member, resolving with the tally after this voice landed.
   *
   * The vote is the members' own recourse, needing no rank: the first call opens the vote, each
   * further member's call adds to it, and when {@link RoomVoteKickResponse.votes} reaches `needed`
   * — half the room, rounded up — the kick lands and `open` turns false. A caller's repeated vote is
   * idempotent. The same tally is broadcast to the room on the {@link onVote} stream, so a client
   * that voted and one that only watched converge on the same count.
   */
  async voteKick(roomId: Id, targetId: Id): Promise<RoomVoteKickResponse> {
    const request: RoomVoteKick = { roomId, targetId };
    return this.#rpc.call(
      OP.ROOM_VOTE_KICK,
      encodeRoomVoteKick,
      decodeRoomVoteKickResponse,
      request,
    );
  }

  /**
   * Applies a moderation action to a member: mute, unmute, kick, ban, or unban.
   *
   * This is the staff path, not the vote: the server admits it only from a caller who outranks the
   * target (a room moderator or above) or a global admin, and rejects it otherwise. `action` is a
   * {@link SanctionAction}; a room mute silences the target for the room (distinct from a caller's
   * personal mute), a kick removes with the door open, a ban bars re-entry, and their undos reverse
   * each. `reason` is an optional note the server may record. The kept-in-sync membership then
   * arrives on the ordinary {@link onMember} stream as a {@link MemberChange}.
   */
  async sanction(args: {
    roomId: Id;
    targetId: Id;
    action: SanctionAction;
    reason?: string;
  }): Promise<Acknowledged> {
    const request: RoomSanction = {
      roomId: args.roomId,
      targetId: args.targetId,
      action: args.action,
    };
    if (args.reason !== undefined) {
      request.reason = args.reason;
    }
    return this.#rpc.call(OP.ROOM_SANCTION, encodeRoomSanction, decodeAcknowledged, request);
  }

  /**
   * Reads a room's roster, highest role first.
   *
   * Each {@link RosterEntry} carries the account, its {@link RoomRole} as a number, and when it
   * joined. `limit` bounds one page; `after` is the cursor — the last `accountId` of the previous
   * page — so a caller pages until a short page arrives. Unlike the live {@link onMember} stream,
   * this is a snapshot for rendering a member list, not a mirror to maintain.
   */
  async getRoster(roomId: Id, limit?: number, after?: Id): Promise<RosterEntry[]> {
    const request: RosterReq = { roomId };
    if (limit !== undefined) {
      request.limit = limit;
    }
    if (after !== undefined) {
      request.after = after;
    }
    const response: RosterResponse = await this.#rpc.call(
      OP.ROOM_ROSTER,
      encodeRosterReq,
      decodeRosterResponse,
      request,
    );
    return response.members;
  }
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Rooms.kt
package com.migo.core.domain

import com.migo.core.protocol.Acknowledged
import com.migo.core.protocol.Op
import com.migo.core.protocol.RoomCreate
import com.migo.core.protocol.RoomJoinRequest
import com.migo.core.protocol.RoomJoinResponse
import com.migo.core.protocol.RoomLeaveRequest
import com.migo.core.protocol.RoomListRequest
import com.migo.core.protocol.RoomListResponse
import com.migo.core.protocol.RoomMemberEvent
import com.migo.core.protocol.RoomSanction
import com.migo.core.protocol.RoomStateEvent
import com.migo.core.protocol.RoomVoteEvent
import com.migo.core.protocol.RoomVoteKick
import com.migo.core.protocol.RoomVoteKickResponse
import com.migo.core.protocol.RosterReq
import com.migo.core.protocol.RosterResponse
import com.migo.core.protocol.SanctionAction
import com.migo.core.wire.Id

/**
 * Rooms: the public and semi-public spaces, as opposed to private conversations.
 *
 * A port of `packages/sdk/src/domains/rooms.ts`. The largest of the small domains, because it is the
 * only one with both requests and three separate event streams -- who is in the room, what the room
 * itself is doing, and how a kick vote against a member is running.
 *
 * # A room is not end-to-end encrypted, and this SDK must not imply that it is
 *
 * Section 178 forbids claiming end-to-end encryption for a Public or Managed Room, and the reason is
 * structural rather than an implementation gap: a room anyone can join, with moderation and history
 * for newcomers, cannot also be a space whose contents only the current members can read. Somebody
 * joining tomorrow gets today's history, which means the history has to be readable by the server that
 * serves it.
 *
 * [RoomJoinResponse.encryption] states the mode the joined conversation actually runs in, and it is
 * there so a client can *show* the difference. Rendering a room with the same lock icon a private chat
 * gets would be the one dishonesty that matters in a messaging app, because a person choosing where to
 * say something is relying on that icon.
 *
 * # Joining hands back a conversation, and the messages come from elsewhere
 *
 * A room's traffic is a conversation like any other: [RoomJoinResponse] carries the
 * [RoomJoinResponse.conversationId] and its [RoomJoinResponse.lastSeq], and from that point messages
 * arrive through [MessagingDomain] and history through [SyncDomain]. There is no room-specific send
 * path, which is what keeps one message pipeline in this SDK instead of two.
 *
 * # Streams because they answer different questions
 *
 * [RoomMemberEvent] is a person arriving, leaving, or changing role -- a discrete thing a UI puts in
 * the timeline. [RoomStateEvent] is the room's own shape: the online count ticking, a topic change, a
 * slow mode going on. Folding them into one stream would make every member list update wake up the
 * code that renders the header, and in a busy room those fire at very different rates.
 *
 * [RoomVoteEvent] is the third and the narrowest: the running tally of a kick vote against one member,
 * coalesced per room so a UI can show "3/17" beside the target while the vote is open and take the row
 * down when it closes. It is separate for the same reason -- a vote is a rare, targeted thing, and a
 * client that never renders a tally simply never adds a handler for it.
 */
class RoomsDomain(
    private val rpc: Rpc,
    onEventError: EventErrorHandler? = null,
) {
    private val memberListeners = ListenerSet<RoomMemberEvent>(Op.ROOM_MEMBER_EVENT, onEventError)
    private val stateListeners = ListenerSet<RoomStateEvent>(Op.ROOM_STATE_EVENT, onEventError)
    private val voteListeners = ListenerSet<RoomVoteEvent>(Op.ROOM_VOTE_EVENT, onEventError)

    @Volatile
    private var subscriptions: List<Subscription>? = null

    /**
     * Begins delivering room events to registered handlers. Idempotent.
     *
     * All three streams are subscribed together: a client that wanted only one of them would still be
     * sent the others by the server, and an unsubscribed opcode reaching [Rpc] is a warning rather than
     * a silent discard.
     */
    fun start() {
        if (subscriptions != null) return
        subscriptions = listOf(
            rpc.on(Op.ROOM_MEMBER_EVENT, { r -> RoomMemberEvent.decode(r) }) { event, _ ->
                memberListeners.deliver(event)
            },
            rpc.on(Op.ROOM_STATE_EVENT, { r -> RoomStateEvent.decode(r) }) { event, _ ->
                stateListeners.deliver(event)
            },
            rpc.on(Op.ROOM_VOTE_EVENT, { r -> RoomVoteEvent.decode(r) }) { event, _ ->
                voteListeners.deliver(event)
            },
        )
    }

    /** Stops delivery of all three streams. Registered handlers are kept for a later [start]. */
    fun stop() {
        val live = subscriptions ?: return
        subscriptions = null
        for (subscription in live) subscription.cancel()
    }

    /**
     * Registers a handler for members arriving, leaving, and changing role.
     *
     * [RoomMemberEvent.memberCount] is the room's total after the change when the server chose to send
     * it, so a client can correct a drifting local count instead of incrementing forever.
     */
    fun onMember(listener: Listener<RoomMemberEvent>): Subscription = memberListeners.add(listener)

    /**
     * Registers a handler for the room's own state.
     *
     * Every field on a [RoomStateEvent] is optional and only what changed is present. An absent field
     * means "unchanged", not "zero" -- treating a missing [RoomStateEvent.onlineCount] as 0 is how a
     * busy room comes to show as empty.
     */
    fun onState(listener: Listener<RoomStateEvent>): Subscription = stateListeners.add(listener)

    /**
     * Registers a handler for the running tally of a kick vote.
     *
     * A [RoomVoteEvent] arrives each time a voice lands on an open vote and once more when it ends.
     * [RoomVoteEvent.closed] tells the two endings apart: true when the vote expired or the target
     * left, absent (null) while the vote is still open, and the kick that a *passed* vote causes
     * arrives separately as a [RoomMemberEvent] with [com.migo.core.protocol.MemberChange.Kicked]. So a
     * UI shows the tally against the target while events keep coming and clears it on any close.
     */
    fun onVote(listener: Listener<RoomVoteEvent>): Subscription = voteListeners.add(listener)

    /**
     * Creates a room and enters it, resolving with the same join handle [join] returns.
     *
     * The caller becomes the room's Owner ([com.migo.core.protocol.RoomRole.Owner]): the one role
     * that can appoint managers and the one a room cannot lose. `slug` is the room's permanent
     * address and must be unique server-side; `kind` picks the governance line —
     * [com.migo.core.protocol.RoomKind.Public] for a community room,
     * [com.migo.core.protocol.RoomKind.Managed] for one under server moderation. The reply is a
     * join response because creation is entry: the creator is the first member.
     */
    suspend fun create(
        slug: String,
        name: String,
        kind: com.migo.core.protocol.RoomKind,
        topic: String? = null,
    ): RoomJoinResponse {
        val request = RoomCreate(slug, name, kind.wire.toLong(), topic)
        return rpc.call(
            Op.ROOM_CREATE,
            { w -> request.encode(w) },
            { r -> RoomJoinResponse.decode(r) },
        )
    }

    /**
     * Joins a room.
     *
     * [inviteCode] is required for a room that is not open to everyone and ignored for one that is.
     * After this returns, subscribe to the returned conversation's topic and sync it from
     * [RoomJoinResponse.lastSeq] to show recent history.
     *
     * Joining a room already joined succeeds and returns the current state rather than failing, so a
     * client reconnecting does not have to track whether it is a member.
     */
    suspend fun join(roomId: Id, inviteCode: String? = null): RoomJoinResponse {
        val request = RoomJoinRequest(roomId, inviteCode)
        return rpc.call(
            Op.ROOM_JOIN,
            { w -> request.encode(w) },
            { r -> RoomJoinResponse.decode(r) },
        )
    }

    /**
     * Leaves a room.
     *
     * The conversation stops delivering, and a client should drop its local membership state on
     * success. Sender-key material for a room conversation is not this domain's to clean up; a client
     * that keeps per-conversation crypto state calls [MessagingDomain.forget] for the conversation
     * afterwards.
     */
    suspend fun leave(roomId: Id): Acknowledged {
        val request = RoomLeaveRequest(roomId)
        return rpc.call(
            Op.ROOM_LEAVE,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }

    /**
     * Browses or searches rooms.
     *
     * [query] searches names and topics; [category], [language] and [country] narrow the directory.
     * All four are independent filters combined with and, and a request with none of them is the plain
     * directory listing. Paginate with [RoomListResponse.nextCursor], exactly as in
     * [ConversationsDomain.list].
     */
    suspend fun list(
        limit: Long,
        query: String? = null,
        category: String? = null,
        language: String? = null,
        country: String? = null,
        cursor: String? = null,
    ): RoomListResponse {
        val request = RoomListRequest(limit, query, category, language, country, cursor)
        return rpc.call(
            Op.ROOM_LIST,
            { w -> request.encode(w) },
            { r -> RoomListResponse.decode(r) },
        )
    }

    /**
     * Reads a room's roster: its members, highest role first.
     *
     * [limit] bounds the page and 0 asks for the server's own page size; [after] is the cursor -- the
     * last member's account id from the previous page. Each [com.migo.core.protocol.RosterEntry]
     * carries the member's role as a raw wire value (a [com.migo.core.protocol.RoomRole] discriminant),
     * because the roster is a bulk listing and a role the server adds tomorrow should page through an
     * old client rather than fail to decode.
     */
    suspend fun getRoster(roomId: Id, limit: Int = 0, after: Id? = null): RosterResponse {
        val request = RosterReq(roomId, if (limit > 0) limit.toLong() else null, after)
        return rpc.call(
            Op.ROOM_ROSTER,
            { w -> request.encode(w) },
            { r -> RosterResponse.decode(r) },
        )
    }

    /**
     * Casts a voice in a kick vote against a member, opening the vote if it is the first.
     *
     * The reply is the tally the moment this voice landed: [RoomVoteKickResponse.votes] of
     * [RoomVoteKickResponse.needed], with [RoomVoteKickResponse.open] false once the vote has passed
     * and the kick already landed. Every other member's voice reaches all watchers as a
     * [RoomVoteEvent], so a client that renders a live tally leans on [onVote] for the rest and uses
     * this reply only to reflect the caller's own vote at once.
     */
    suspend fun voteKick(roomId: Id, targetId: Id): RoomVoteKickResponse {
        val request = RoomVoteKick(roomId, targetId)
        return rpc.call(
            Op.ROOM_VOTE_KICK,
            { w -> request.encode(w) },
            { r -> RoomVoteKickResponse.decode(r) },
        )
    }

    /**
     * Applies a moderation action to a member: mute, unmute, kick, ban, or unban.
     *
     * A staff power rather than a vote: the server enforces that the caller outranks the target, so a
     * client should only offer the actions a [com.migo.core.protocol.RoomRole] of Moderator or above
     * holds over a strictly lower role. [reason] is optional and, where a room keeps a moderation log,
     * is what gets written to it. The reply is a bare acknowledgement, so nothing is returned here --
     * the membership change the action causes arrives on its own as a [RoomMemberEvent], and that is
     * what a UI reacts to.
     */
    suspend fun sanction(
        roomId: Id,
        targetId: Id,
        action: SanctionAction,
        reason: String? = null,
    ) {
        val request = RoomSanction(roomId, targetId, action, reason)
        rpc.call(
            Op.ROOM_SANCTION,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }
}
